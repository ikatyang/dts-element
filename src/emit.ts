import * as ts from 'typescript';
import { IElement } from './element';
import { transform } from './transform';

export const emit_native = (node: ts.Node, options?: ts.PrinterOptions) => {
  const source_file = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(options);
  return printer.printNode(ts.EmitHint.Unspecified, node, source_file);
};

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) =>
  emit_native(transform(element), options);
