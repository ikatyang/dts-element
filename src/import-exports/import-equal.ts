import * as ts from 'typescript';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';

export interface IImportEqualOptions extends IElementOptions {
  from: string;
  name: string;
}

export interface IImportEqual
  extends IElement<ElementKind.ImportEqual>,
    IImportEqualOptions {}

export const create_import_equal = (
  options: IImportEqualOptions,
): IImportEqual => ({
  ...create_element(ElementKind.ImportEqual),
  ...options,
});

export const is_import_equal = (value: any): value is IImportEqual =>
  is_element(value) && value.kind === ElementKind.ImportEqual;

// tslint:disable:ter-indent

/**
 * @hidden
 */
export const transform_import_equal = (
  element: IImportEqual,
  _path: IElement<any>[],
) =>
  ts.createImportEqualsDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ undefined,
    /* name            */ ts.createIdentifier(element.name),
    /* moduleReference */ ts.createExternalModuleReference(
      ts.createLiteral(element.from),
    ),
  );
