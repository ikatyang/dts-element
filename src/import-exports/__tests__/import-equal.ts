import { emit } from '../../emit';
import { create_import_equal, is_import_equal } from '../import-equal';

it('should return correctly with name and from', () => {
  expect(
    emit(
      create_import_equal({
        name: 'abc',
        from: 'xyz',
      }),
    ),
  ).toMatchSnapshot();
});

describe('is_import_equal', () => {
  it('should return correctly', () => {
    const element = create_import_equal({
      name: 'abc',
      from: 'xyz',
    });
    expect(is_import_equal(element)).toBe(true);
  });
});
