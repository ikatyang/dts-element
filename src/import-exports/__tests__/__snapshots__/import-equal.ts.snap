// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with name and from 1`] = `"import abc = require(\\"xyz\\");"`;
